app/src/main/java/ca/marshallasch/veil/DataStore.java

```diff
@@ -272,4 +285,92 @@ public void syncData(Sync.HashData message) {
         }
     }
 
+    /**
+     * This will generate a data synchronization message for the given peer.
+     * This is the version 2 of the data synchronization protocol.
+     * @param peer the {@link MeshId} for the peer to send the sync message too
+     * @return a sync message that is filled with the data for that peer
+     */
+    @NonNull
+    public Sync.SyncMessage getSyncFor(MeshId peer) {
+
+        // get time last sent data
+        Sync.SyncMessage.Builder builder = Sync.SyncMessage.newBuilder();
+
+        Date timeLastSentData = db.getTimeLastSentData(peer.toString());
+
+        // get the list of comments and post hashes since the given time.
+        List<Pair<String, String>> mapping = db.dumpKnownPosts(timeLastSentData);
+
+        Set<String> hashes = new ArraySet<>();
+
+        for (Pair<String, String> pair: mapping) {
+            hashes.add(pair.first);
+            hashes.add(pair.second);
+
+            builder.addMappings(Sync.CommentMapping.newBuilder()
+                    .setPostHash(pair.first)
+                    .setCommentHash(pair.second)
+                    .build());
+        }
+        // remove the empty string from the empty comment hashes
+        hashes.remove("");
+
+        DhtProto.DhtWrapper wrapper;
+
+        for (String hash: hashes) {
+
+            // search for the comment or post
+            wrapper = hashTableStore.getPostOrComment(hash);
+
+            // insert into the list
+            if (wrapper != null) {
+                builder.addEntries(Sync.HashPair.newBuilder()
+                        .setHash(hash)
+                        .setEntry(wrapper)
+                        .build());
+            }
+        }
+
+        return builder.build();
+    }
+
+    /**
+     * This will insert the data sync message into the data store and the database.
+     * This is for the version 2 message.
+     *
+     * @param syncMessage the message from the remote peer to save.
+     */
+    public void insertSync(@Nullable Sync.SyncMessage syncMessage) {
+
+        if (syncMessage == null) {
```

Comment:
why not just make Sync.SyncMessage @NonNullable

Reply:
this way you don't have to do null checks in the caller, and the function will take care of it.